Projects/Features/PlaylistFeature/Sources/Reactors/PlaylistDetailReactor.swift

```diff
@@ -0,0 +1,215 @@
+        case .save:
+            newState.isEditing = false
+        case let .changeSelectedState((data, count)), let .changeAllState((data, count)):
+            newState.dataSource = [PlayListDetailSectionModel(model: 0, items: data)]
+            newState.selectedItemCount = count
+        case .undo:
+            newState.dataSource = state.backupDataSource
+            newState.isEditing = false
+        }
+
+        return newState
+    }
+}
+
+// MARK: - Mutate
+
+private extension PlaylistDetailReactor {
+    /// 서버에서 데이터 불러오기
+    func fetchData() -> Observable<Mutation> {
+        return fetchPlayListDetailUseCase.execute(id: key, type: type)
+            .catchAndReturn(
+                PlayListDetailEntity(
+                    key: key,
+                    title: "",
+                    songs: [],
+                    image: "",
+                    image_square_version: 0,
+                    image_round_version: 0,
+                    version: 0
+                )
+            )
+            .asObservable()
+            .map { [weak self] result in
+
+                guard let self else {
+                    return PlaylistMetaData(
+                        list: [],
+                        header: PlayListHeader(title: "", songCount: "", image: "", version: 0)
+                    )
+                }
+                return PlaylistMetaData(
+                    list: [PlayListDetailSectionModel(model: 0, items: result.songs)],
+                    header: PlayListHeader(
+                        title: result.title,
+                        songCount: "\(result.songs.count)곡",
+                        image: self.type == .wmRecommend ? result.key : result.image,
+                        version: self.type == .wmRecommend ? result.image_square_version : result.version
+                    )
+                )
+            }
+            .map(Mutation.fetchData)
+    }
+
+    /// 순서 변경
+    func updateOrder(src: Int, dest: Int) -> Observable<Mutation> {
+        var tmp = (currentState.dataSource.first ?? PlayListDetailSectionModel(model: 0, items: [])).items
+        let target = tmp[src]
+        tmp.remove(at: src)
+        tmp.insert(target, at: dest)
+        return .just(.updateOrder(tmp))
+    }
+
+    /// 저장(서버)
+    func saveData() -> Observable<Mutation> {
+        let dataSource = currentState.dataSource[0].items.map { $0.id }
+        let backupDataSource = currentState.backupDataSource[0].items.map { $0.id }
+
+        if dataSource.elementsEqual(backupDataSource) {
+            return .empty()
+        }
+
+        return editPlayListUseCase
+            .execute(key: key, songs: dataSource)
+            .asObservable()
+            .map { _ in .save }
+
+        // 여기서 새로운 데이터 패치 해야하는데 ??
+    }
+
+    /// 단일 곡 선택 상태 변경
+    func changeSelectingState(_ index: Int) -> Observable<Mutation> {
+        var tmp = (currentState.dataSource.first ?? PlayListDetailSectionModel(model: 0, items: [])).items
```

Comment:
```suggestion
        guard var tmp = currentState.dataSource.first?.items else { return .empty() } 
```
이 친구 dataSource가 nil이여서 기본 PlayListDetailSectionModel 들어가면 아래 에서 tmp[index] 에서 크래시 터질거같은데 그냥 guard 언래핑하는거 어떤가요?
겸사겸사 가독성도 더 깔끔해질거같네요

Reply:
return .empty()하기 전에 LogManager.printError("playlist detail datasource is empty")
대충 이런 로그 찍는거 있어도 나중에 혹여나 트래킹할때 좋을거같아요

Reply:
적용